Validate input against a template; closes #2.

bast opened the pull request:
Prototyping input validation. For reference see #2.

Few notes:
- Sources are not in the right place yet, I will split and place them in the right place. Now I just wanted to give you the chance to have a look and also I wanted to back up my work before leaving office.
- Tests to be written and hooked up with rest of the testing, coming soon.
- Complex numbers do not work yet because YAML reader turns them into a string so we will need a workaround somewhere - I postponed this.

I know that the Python can be beautified and compacted but already at this stage all feedback welcome. You can run the example by stepping into the directory and running `validate.py`.

bast committed:
prototyping input validation

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

robertodr commented:
As for complex numbers, I will remove them from the Getkw grammar, as they were not part of the original "specification". We might have to resort to separate real and imaginary parts or have an implicit translation predicate from `str` to `complex`.

bast commented:
So at least from my (validation) perspective, complex numbers are not a problem - I was only "lazy" to solve it today but it's a two-line workaround. But perhaps I misunderstood the comment.

robertodr commented on validation/validate.py:
```diff
@@ -0,0 +1,215 @@
+
+    if expected_type_is_list:
+        # make sure that value is actually a list
+        if not type(value).__name__ == 'list':
```
Can `isinstance` be used instead?

bast replied:
Yes, I think `isinstance` can be used - I will adjust.

bast replied:
I will first move/split the files, enable testing, and then I will replace with the `isinstance`. So it's not forgotten but I don't want to change this blindly - yesterday I was testing interactively, now I first want to set up automated tests for the many checks before touching the code.

bast replied:
Aha - now I remember why I did not use isinstance: I wanted to compare the string:
```python
# wrong but would be convenient for me
isinstance(3, 'int')

# correct but inconvenient for me
isinstance(3, int)
```
I am sure this can be somehow solved but how about I place a comment into the code and I or somebody else can be later creative about it?

robertodr replied:
Go for the comment! Probably we need to put a `lambda` there that returns the type given the string.

robertodr commented on validation/validate.py:
```diff
@@ -0,0 +1,215 @@
+
+        # iterate over each element of the list
+        # and check whether it matches T
+        list_element_type = re.search(r"^List\[(\w+)\]$", expected_type).group(1)
```
Isn't homogeneity of the collection ensured if its type is `list`? Probably this check is redundant?

bast replied:
Sadly no:
```
>>> l = [1, 1.0, 'foo']
>>> type(l).__name__
'list'
```

robertodr replied:
Wow! Though not so surprising given that Python doesn't have a "real" type system...

robertodr commented on validation/validate.py:
```diff
@@ -0,0 +1,215 @@
+
+def validate_node(input_dict, template_dict):
+    '''
+    Verify a node. A node is either the root of the input or it is a section.
```
For the docstrings, we should adhere to the NumPy style guide: https://numpydoc.readthedocs.io/en/latest/format.html

bast replied:
I agree, but we can adjust this after the code is in.

bast replied:
What I meant by this: the code will still change quite a bit, probably even in this PR so I did not want to make the docstrings too beautiful.

robertodr replied:
Totally understand, but I'd prefer if the un-WIP form of this PR had docstrings in the "right" format :smiling_imp:

bast replied:
OK - first rough stone masonry. Later cosmetics.

robertodr commented on validation/validate.py:
```diff
@@ -0,0 +1,215 @@
+
+    Can raise either InputError or TemplateError.
+    '''
+    input_sections = [x for x in input_dict if type(input_dict[x]).__name__ == 'dict']
```
I would use `isinstance` rather than `type(x).__name__ ==`

bast replied:
I agree - I will adjust.

bast replied:
Same as above - resolving this but I keep the above conversation.

robertodr reviewed (commented):
I mostly have comments about "beautifying" the use of Python. The logic and the implementation are as we discussed in the issue and over the chat. Great job!

robertodr commented on validation/validate.py:
```diff
@@ -0,0 +1,215 @@
+
+
+if __name__ == '__main__':
+    input_dict = read_yaml_file('example.yml')
```
When this consolidates with code and tests in the right directories, we'll need to add tests for directly passing a `dict` inside and using a JSON file.

bast replied:
Agreed - this is the plan.

bast replied:
But I still might start with file-based tests until the concrete solidifies a bit. The reason is that it is easier to change files rather than change/edit nested dictionaries.

robertodr replied:
The testing with `dict`-s can come later. I have to investigate how to automate it/randomize it with Hypothesis and then we can have a one (or few) functions generating test cases as `dict`-s with serialization to JSON/YAML done separately.

bast replied:
Actually I now realized that I will need some dicts for the tests. I will add a number of hand-written tests, later we can make it more sophisticated with hypothesis.

bast replied:
Can I resolve this discussion or do you want to follow up with an issue?

robertodr replied:
No need to follow up with an issue.

bast committed:
add test_type_matches to the test machinery

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
run the validation as a test

up to now there is not much being tested, we only test
that the code does not crash

more fine-grained testing coming up

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
pathlib-workaround for python < 3.6

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
compare the result input dict with reference

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
fix type hint for JSON

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
mv read_yaml_file under parselglossy/

since we will need this not only for testing

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
nicer docstrings

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
use a helper function for tests

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
test expected template and input errors

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
run pytest with -s so that we can see debug prints if needed

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

robertodr commented on .travis.yml:
```diff
@@ -18,7 +18,7 @@ install:
 script:
   - pipenv run flake8 parselglossy
-  - pipenv run py.test -v --cov=parselglossy tests/ --doctest-module
+  - pipenv run py.test -s -v --cov=parselglossy tests/ --doctest-module
```
:+1: for adding the flag. Don't forget to add it also to `.ci/azure-steps.yml`

bast replied:
Done.

robertodr commented on parselglossy/validate.py:
```diff
@@ -0,0 +1,195 @@
+def type_matches(value: Any, expected_type: str) -> bool:
+    """Checks whether value has the type expected_type.
+
+    If yes, returns True, otherwise False.
```
Can we have this under a "Returns" block?

robertodr commented on parselglossy/validate.py:
```diff
@@ -0,0 +1,195 @@
+
+    If yes, returns True, otherwise False.
+
+    Allowed types T are: 'str', 'int', 'float', 'complex', 'bool',
```
This one under a "Notes" block. Probably the types should be in backticks too, so they render properly.

robertodr commented on parselglossy/validate.py:
```diff
@@ -0,0 +1,195 @@
+from typing import Any, Dict
+
+
+JsonDict = Dict[str, Any]
```
Nitpick: can we name the type `JSONDict`?

bast replied:
I took Guido's naming: https://github.com/python/typing/issues/182#issuecomment-185996450 - but no problem for me to change it.

robertodr commented on Pipfile:
```diff
@@ -21,6 +21,7 @@ twine = "*"
 watchdog = "*"
 wheel = "*"
 yapf = "*"
+pyyaml = "*"
```
Shouldn't this go under `[packages]` and also in requirement in `setup.py`?

robertodr commented on parselglossy/validate.py:
```diff
@@ -0,0 +1,195 @@
+    template_dict: JsonDict
+        Contains the input template which holds the predicates.
+
+    Returns
```
The `Returns` block can be removed altogether.

robertodr commented on parselglossy/validate.py:
```diff
@@ -0,0 +1,195 @@
+
+def check_predicates_node(input_dict: JsonDict,
+                          input_dict_node: JsonDict,
+                          template_dict_node: JsonDict) -> None:
```
This function raises as soon as the first predicate fails, am I correct? As a final user I would prefer to know everything that's wrong with my input file, as the back and forth is quite unpleasant. Follow up and discussion in #24

robertodr commented on tests/test_validation.py:
```diff
@@ -0,0 +1,141 @@
+from typing import Dict, Any
+
+
+JsonDict = Dict[str, Any]
```
Can we put the type hint in `atoms.py` and then import it from there?

robertodr commented on tests/test_validation.py:
```diff
@@ -0,0 +1,141 @@
+def _helper(category: str,
+            input_file_name: str,
+            template_file_name: str) -> JsonDict:
+    this_path = Path(os.path.dirname(os.path.realpath(__file__)))
```
`Path(__file__)` is enough.

bast replied:
Thanks for this - this is a major simplification and one more reason to go for `pathlib`.

bast replied:
No - this was not correct since I need the directory, not the path. We can fix this later but for the moment I reverted my change.

robertodr replied:
Agreed, probably it's `Path(__file__).parent` or some such.

robertodr commented on tests/test_validation.py:
```diff
@@ -0,0 +1,141 @@
+                 'energy': 1e-05
+             },
+            'another_number': 10,
+            'some_complex_number': '0.0+0.0j'
```
Why is the complex number a string? Can you add a comment here and open a follow-up issue if this is a known limitation?

bast replied:
I will follow up with issue. This is what the YAML reader returns we will need to cast it into a complex number.

robertodr replied:
Thanks!

robertodr commented on tests/test_validation.py:
```diff
@@ -0,0 +1,141 @@
+        input_dict = _helper('template_errors',
+                             'input.yml',
+                             'template_no_documentation.yml')
+    assert str(e.value) == "keyword(s) without any documentation: ['a_short_string']"
```
Is the cast to string needed?

bast replied:
Yes - otherwise we get more than the error string. Which is also possible but here I only wanted the error string.

robertodr commented on tests/test_validation.py:
```diff
@@ -0,0 +1,141 @@
+    assert str(e.value) == "the following keyword(s) must be set: {'a_short_string'}"
+
+    # type errors
+    for file_name, error in [
```
I think the test can be refactored using `parametrize` rather than `for` loops internally. Would you mind adding a note to #25?

robertodr commented on tests/test_validation.py:
```diff
@@ -0,0 +1,141 @@
+        input_dict = _helper('input_errors',
+                             'input_predicate_cross.yml',
+                             'template.yml')
+    # in python < 3.6 the dict order is not guaranteed so we are not sure
```
So in Py3.6+ `dict`-s are ordered by default? Good to know!

bast replied:
Yes, order is preserved. This is very nice.

robertodr commented:
Really looking awesome! Thanks @bast :confetti_ball: :green_apple:

bast committed:
s/JsonDict/JSONDict/g

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
mv comment under "Returns" block

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
mv comment under "Notes" block

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
rm a "Returns" block

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
require pyyaml in setup.py

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
add comment about isinstance vs type(x).__name__

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>

bast committed:
add comment about problem with complex numbers

they are currently parsed as strings in the validator

Signed-off-by: Radovan Bast <bast@users.noreply.github.com>
